lib/spack/spack/package.py

```diff
@@ -1559,60 +1570,65 @@ def run_test(self, exe, options=[], expected=[], status=None,
             except FileNotFoundError:
                 self._binaries = []
 
-        if skip_missing and exe not in self._binaries:
-            return
+        if skip_missing:
```

Comment:
We need to run `which(exe)` at some point to check whether it's missing, since it could be something we expect to find from the test environment.

We could modify `_run_test_helper` to accept an `Executable` instead of a string, and then in `run_test` we could run `exec = which(exe, required=not skip_missing)` to raise an error if we can't find the executable.

Reply:
> We need to run `which(exe)` at some point to check whether it's missing, since it could be something we expect to find from the test environment.

Okay.  That does change the meaning of the `skip_missing` option, which was originally for skipping missing installed `bin` files.  (The intention was to avoid tracking all of the installed binaries expected for each version of a package.)

> 
> We could modify `_run_test_helper` to accept an `Executable` instead of a string, and then in `run_test` we could run `exec = which(exe, required=not skip_missing)` to raise an error if we can't find the executable.

Okay.  I'll pursue this approach.

Reply:
I forgot to mention last Friday that I finished a pass at these changes last Friday (May 8th).